rclcpp/src/rclcpp/guard_condition.cpp

```diff
@@ -96,8 +99,6 @@ GuardCondition::exchange_in_use_by_wait_set_state(bool in_use_state)
 void
 GuardCondition::add_to_wait_set(rcl_wait_set_t * wait_set)
 {
-  std::lock_guard<std::recursive_mutex> lock(reentrant_mutex_);
-
```

Comment:
How about the following MT case?

1. Thread-A comes with `wait_set-X`.
2. Context switch to thread-B right after thread-A stores `wait_set-X` to `wait_set_`.
3. Thread-B also comes with `wait_set-X`.
4. Thread-B and Thread-A might call `rcl_wait_set_add_guard_condition` at the same time.

as far as i see the doc of `rcl_wait_set_add_subscription`, these functions are not thread-safe?

Reply:
I'm not sure.
The other waitables (e.g. actions, intra-process sub) do not use a mutex there.
Also the executor does not seem to use a mutex when adding entities (subscriptions, services, etc) to waitsets.

Reply:
It's undefined behavior to add an item to two different wait sets, and it's also undefined behavior to try to wait on a single wait set from two different threads.

There is a thread-safe `rclcpp::WaitSet` that protects this case from happening.

Reply:
It looks to me that the `add_to_wait_set`  function is already preventing to add an item to two different wait-sets.

This is the current code.
The atomic `in_use_by_wait_set_` should be enough to avoid the undefined behavior, without need for a mutex.

```
void
GuardCondition::add_to_wait_set(rcl_wait_set_t * wait_set)
{
  if (exchange_in_use_by_wait_set_state(true)) {
    if (wait_set != wait_set_) {
      throw std::runtime_error("guard condition has already been added to a wait set.");
    }
  } else {
    wait_set_ = wait_set;
  }

  rcl_ret_t ret = rcl_wait_set_add_guard_condition(wait_set, &this->rcl_guard_condition_, NULL);
  if (RCL_RET_OK != ret) {
    rclcpp::exceptions::throw_from_rcl_error(
      ret, "failed to add guard condition to wait set");
  }
}
```

Reply:
It looks to me that the `add_to_wait_set`  function is already preventing to add an item to two different wait-sets.

This is the current code.
The atomic `in_use_by_wait_set_` should be enough to avoid the undefined behavior, without need for a mutex.

```
void
GuardCondition::add_to_wait_set(rcl_wait_set_t * wait_set)
{
  if (exchange_in_use_by_wait_set_state(true)) {
    if (wait_set != wait_set_) {
      throw std::runtime_error("guard condition has already been added to a wait set.");
    }
  } else {
    wait_set_ = wait_set;
  }

  rcl_ret_t ret = rcl_wait_set_add_guard_condition(wait_set, &this->rcl_guard_condition_, NULL);
  if (RCL_RET_OK != ret) {
    rclcpp::exceptions::throw_from_rcl_error(
      ret, "failed to add guard condition to wait set");
  }
}
```